scanpy/tools/_umap.py

```diff
@@ -131,7 +132,9 @@ def umap(
         n_epochs,
         init_coords,
         random_state,
-        max(0, verbosity-3))
+        metric="euclidean",
```

Comment:
I don't think that a metric makes any sense here. We're just embedding the graph that had already been computed before. Are you sure that `simplicial_set_embedding` still takes a metric as an argument?

Reply:
Yes, unfortunately it does. Should i add metric as an argument to `umap`.

Reply:
Ok, just looked it up, it's just used for the initialization.

Same as the `data` argument, which, btw, is not correct to be set to `adata.X`. It should be set to the representation chosen in `pp.neighbors` and the metric also needs to be chosen consistently with it.

I'd recommend to stare the params of the `neighbors` call as `.uns['neighbors']['params']` as a dict as in several other tools, too and initizalize from that.

Reply:
Hm, in the [description ](https://github.com/lmcinnes/umap/blob/a858c6322a3e682d8daf9c17e13ac023f3e18cfa/umap/umap_.py#L856) it says

> data: array of shape (n_samples, n_features)
> The source data to be embedded by UMAP

It is used in [spectral_layout](https://github.com/lmcinnes/umap/blob/9342e43ad56cb93e63089997fb84e6484f8b1f24/umap/spectral.py#L199) only, it uses data and metric only when the number of connected components is bigger than 1. And it also seems from [the code](https://github.com/lmcinnes/umap/blob/9342e43ad56cb93e63089997fb84e6484f8b1f24/umap/spectral.py#L49) that this is exactly `adata.X`. Or maybe not `adata.X`, but the data which was used for embedding, `adata.X` could be changed already from that state. So what can we do, froze `adata.X` which were used for embedding?

Reply:
What we should do is this
> I'd recommend to store the params of the `pp.neighbors` call as `.uns['neighbors']['params']` as a `dict` as in several other tools, too and initialize from that.

So, you'll have a `use_rep` param in that dict. And that param tells you which representation intends to use for the embedding (used for computing the neighborhood graph/simplicial set).

Reply:
Yes, thank you, i see . Will do like this.

Reply:
Just forgot that there can be another representation to use.